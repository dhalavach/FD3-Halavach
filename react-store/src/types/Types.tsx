import store from "./../components/Store";

export type Product = {
  id: number;
  className: string;
  itemType: string;
  itemName: string;
  itemPrice: number;
  itemImage: string;
  itemImageURL: string;
  itemImageAlt: string;
  itemQuantity: number;
  itemDescription: string;
  count: number;
};

export type orderObject = {
  id: number;
  name: string;
  email: string;
  address: string;
  orderedProducts: Product[];
};

export type Order = {};

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
