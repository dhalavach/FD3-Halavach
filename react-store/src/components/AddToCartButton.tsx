import { useEffect, useState, ReactElement } from "react";
import { Product, RootState } from "../types/Types";
import { useSelector } from "react-redux";

export default function AddToCartButton(props: {
  add: (product: Product) => void;
  product: Product;
}) {
  const { add, product } = props;
  const [inCart, setInCart] = useState(false);
  const cartProducts = useSelector((state: RootState) => state.cartProducts);
  const checkIfInCart = (product: Product): boolean => {
    return cartProducts.findIndex((p: Product) => p.id === product.id) !== -1;
  };

  useEffect(() => {
    setInCart(checkIfInCart(product));
  }, [cartProducts]);

  return (
    <button
      className={`btn ${
        inCart ? "table__button-in-cart" : "table__button-add"
      }`}
      onClick={() => add(product)}
      data-testid="add-to-cart-button"
    >
      {inCart ? "Product in Cart" : "Add to Cart"}
    </button>
  );
}
