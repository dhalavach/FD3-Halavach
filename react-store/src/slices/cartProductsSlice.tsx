import { createSlice } from "@reduxjs/toolkit";
import { useSelector, useDispatch } from "react-redux";
//import { clients } from './selectedClientSlice';

export const cartProductsSlice = createSlice({
  name: "cartProducts",
  initialState: localStorage.getItem("cartProducts")
    ? JSON.parse(localStorage.getItem("cartProducts") as string)
    : [],

  reducers: {
    setCartProducts: (state, param) => {
      const { payload } = param;
      return payload;
    },
  },
});

export const { setCartProducts } = cartProductsSlice.actions;
export default cartProductsSlice.reducer;
