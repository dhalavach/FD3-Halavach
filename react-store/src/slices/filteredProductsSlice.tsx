import { createSlice } from "@reduxjs/toolkit";

export const filteredProductsSlice = createSlice({
  name: "filteredProducts",
  initialState: [],

  reducers: {
    setFilteredProducts: (state, param) => {
      const { payload } = param;
      return payload;
    },
  },
});

export const { setFilteredProducts } = filteredProductsSlice.actions;
export default filteredProductsSlice.reducer;
