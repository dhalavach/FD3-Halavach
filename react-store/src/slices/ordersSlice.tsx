import { createSlice } from '@reduxjs/toolkit';

const arr: any[] = [];


export const ordersSlice = createSlice({
  name: 'orders',
  initialState: {ordersArr: arr},

  reducers: {
    setOrders: (state, action) => {
      state.ordersArr = [...state.ordersArr, action.payload];
    },
  },
});

export const { setOrders } = ordersSlice.actions;
export default ordersSlice.reducer;
