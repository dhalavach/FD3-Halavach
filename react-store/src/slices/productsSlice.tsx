import { createSlice } from "@reduxjs/toolkit";
import { useSelector, useDispatch } from "react-redux";
//import { clients } from './selectedClientSlice';

export const productsSlice = createSlice({
  name: "products",
  initialState: [],

  reducers: {
    setProducts: (state, param) => {
      const { payload } = param;
      return payload;
    },
  },
});

export const { setProducts } = productsSlice.actions;
export default productsSlice.reducer;
