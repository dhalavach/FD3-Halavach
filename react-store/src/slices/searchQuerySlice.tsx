import { createSlice } from '@reduxjs/toolkit';

export const searchQuerySlice = createSlice({
  name: 'searchQuery',
  initialState: '',

  reducers: {
    setSearchQuery: (state, param) => {
      const { payload } = param;
      return payload;
    },
  },
});

export const { setSearchQuery } = searchQuerySlice.actions;
export default searchQuerySlice.reducer;
