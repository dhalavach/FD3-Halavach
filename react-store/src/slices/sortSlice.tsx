import { createSlice } from "@reduxjs/toolkit";

export const sortSlice = createSlice({
  name: "sort",
  initialState: "",

  reducers: {
    setSort: (state, param) => {
      const { payload } = param;
      return payload;
    },
  },
});

export const { setSort } = sortSlice.actions;
export default sortSlice.reducer;
