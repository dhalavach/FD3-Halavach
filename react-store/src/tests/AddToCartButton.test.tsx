import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import AddToCartButton from '../components/AddToCartButton';
import { Provider } from 'react-redux';
import store from '../components/Store';
import { setCartProducts } from '../slices/cartProductsSlice';
import { useSelector, useDispatch } from 'react-redux';
import { Product } from '../types/Types';

afterEach(cleanup);
const mockProduct = {
  id: 1,
  className: 'Item',
  itemType: 'motherboard',
  itemName: 'MSI MPG B650 Edge WiFi',
  itemPrice: 1095,
  itemImage: '/assets/images/msib650.jpg',
  itemImageURL: 'https://uni.by/files/ItemsImages/137010_r300.jpg',
  itemImageAlt: 'MSI B650 Edge',
  itemQuantity: 0,
  itemDescription: '',
  count: 1,
};
const mockCartProducts = [
  {
    id: 1,
    className: 'Item',
    itemType: 'motherboard',
    itemName: 'MSI MPG B650 Edge WiFi',
    itemPrice: 1095,
    itemImage: '/assets/images/msib650.jpg',
    itemImageURL: 'https://uni.by/files/ItemsImages/137010_r300.jpg',
    itemImageAlt: 'MSI B650 Edge',
    itemQuantity: 0,
    itemDescription: '',
    count: 1,
  },
];

test('component matches the snapshot', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <AddToCartButton />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('button renders without throwing errors', () => {
  // render(
  //   <Provider store={store}>
  //     <AddToCartButton product={mockProduct} />
  //   </Provider>
  // );

  // const button = screen.getAllByTestId('add-to-cart-button')[0];
  // expect(button).toBeTruthy();

  let comp = renderer.create(
    <Provider store={store}>
      <AddToCartButton />
    </Provider>
  );
  expect(comp).toBeTruthy();
});
