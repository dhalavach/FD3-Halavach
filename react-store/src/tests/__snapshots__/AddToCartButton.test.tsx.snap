// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component matches the snapshot 1`] = `
<button
  className="btn table__button-add"
  data-testid="add-to-cart-button"
  onClick={[Function]}
>
  Add to Cart
</button>
`;
